traitsui/testing/ui_tester.py

```diff
@@ -0,0 +1,422 @@
+
+        Parameters
+        ----------
+        object : HasTraits
+            An instance of HasTraits for which a GUI will be created.
+        ui_kwargs : dict or None, optional
+            Keyword arguments to be provided to ``HasTraits.edit_traits``.
+            Default is to call ``edit_traits`` with no additional keyword
+            arguments.
+
+        Yields
+        ------
+        ui : traitsui.ui.UI
+        """
+        return _create_ui(object, ui_kwargs)
+
+    def find_by_name(self, ui, name):
+        """ Find the UI editor with the given name and return an object for
+        simulating user interactions with the editor. The list of
+        ``InteractionRegistry`` in this tester is used for finding the editor
+        specified.
+
+        Parameters
+        ----------
+        ui : traitsui.ui.UI
+            The UI created, e.g. by ``create_ui``.
+        name : str
+            A single or an extended name for retreiving an editor on a UI.
+            e.g. "attr", "model.attr1.attr2"
+
+        Returns
+        -------
+        wrapper : BaseUserInteractor
+        """
+        return (
+            self._get_ui_interactor(ui).locate(locator.TargetByName(name=name))
+        )
+
+    def find_by_id(self, ui, id):
+        """ Find the UI editor with the given identifier and return an object
+        for simulating user interactions with the editor. The list of
+        ``InteractionRegistry`` in this tester is used for finding the editor
+        specified.
+
+        Parameters
+        ----------
+        ui : traitsui.ui.UI
+            The UI created, e.g. by ``create_ui``.
+        id : str
+            Id for finding an item in the UI.
+
+        Returns
+        -------
+        wrapper : BaseUserInteractor
+        """
+        return self._get_ui_interactor(ui).locate(locator.TargetById(id=id))
+
+    def _get_ui_interactor(self, ui):
+        return UIWrapper(
+            editor=ui,
+            registries=self._registries,
+            delay=self.delay,
+        )
+
+
+class UIWrapper:
+    """
+    An ``UIWrapper`` is responsible for dispatching specified user
+    interactions to toolkit specific implementations in order to test a GUI
+    component written using TraitsUI.
+
+    Each instance of ``UIWrapper`` wraps an instance of Editor obtained
+    from a UI object. The editor should be in the state where the GUI has been
+    initialized and has not been disposed of such that it accepts user
+    commands.
+
+    A ``UIWrapper`` finds toolkit specific implementations for simulating
+    user interactions from one or many ``InteractionRegistry`` objects. These
+    registries are typically provided via another public interface, e.g.
+    ``UITester``.
+
+    Attributes
+    ----------
+    editor : Editor
+        An instance of Editor. It is assumed to be in a state after the UI
+        has been initialized but before it is disposed of.
```

Comment:
The `editor` attribute will be renamed, and it may not be an instance of traitsui Editor.
#1107 will clarify that.